### ISSUE TITLE
[YAML] flag to keep refs between calls to parse()

### ISSUE THREAD

**Issue by Doqnach** [symfony/symfony#30343]
**Description**  
For structuring Yaml files we want to use a custom include tag (in our case `!inc/file 'my_file.yml'`). This is working just fine by having a custom `YamlFileLoader` that recursively parses any `TaggedValue` and having it make a `new Parser()` to handle the sub-document (or `self::load()` on the sub-document).

```php
$loaderResolver = new LoaderResolver([new YamlFileLoader()]);
$delegatingLoader = new DelegatingLoader($loaderResolver);
$config = $delegatingLoader->load('my-base.yml');
```

But if your parent document has references, e.g.: `&foo` and you want to reference that within an included file with `*foo`, the sub-parser has no access to the refs from the main document.

At least two ways to solve it:
1. a way to pass refs to a parser
2. don't make a new parser, but calling `parse()` on the same instance as the parent document

Here option 2 seems a nice solution, but: `parse()` sets `$this->refs = []` before and after every call to `doParse()`.

The feature request is to either support option 1, or a flag (e.g.: `Yaml::PARSE_KEEP_REFS`) that disables setting `$this->refs = []` before and after every call to `doParse()`.

**Example**  

```php
const PARSE_KEEP_REFS = 2048;

/**
 * Parses a YAML string to a PHP value.
 *
 * @param string $value A YAML string
 * @param int    $flags A bit field of PARSE_* constants to customize the YAML parser behavior
 *
 * @return mixed A PHP value
 *
 * @throws ParseException If the YAML is not valid
 */
public function parse(string $value, int $flags = 0)
{
    if (false === preg_match('//u', $value)) {
        throw new ParseException('The YAML value does not appear to be valid UTF-8.', -1, null, $this->filename);
    }

    if (!(Yaml::PARSE_KEEP_REFS & $flags)) {
        $this->refs = [];
    }

    $mbEncoding = null;
    $data = null;

    if (2 /* MB_OVERLOAD_STRING */ & (int) ini_get('mbstring.func_overload')) {
        $mbEncoding = mb_internal_encoding();
        mb_internal_encoding('UTF-8');
    }

    try {
        $data = $this->doParse($value, $flags);
    } finally {
        if (null !== $mbEncoding) {
            mb_internal_encoding($mbEncoding);
        }
        $this->lines = [];
        $this->currentLine = '';
        if (!(Yaml::PARSE_KEEP_REFS & $flags)) {
            $this->refs = [];
        }
        $this->skippedLineNumbers = [];
        $this->locallySkippedLineNumbers = [];
    }

    return $data;
}
```
```php
$configValues = $parser->parseFile($resource, Yaml::PARSE_CUSTOM_TAGS|Yaml::PARSE_KEEP_REFS);
```

Of course it's very possible there is a completely other solution that I missed. Since pretty much everything is private, extending the class with your own parser brings too much to the table.

PS: using Symfony config import (https://symfony.com/doc/current/configuration/configuration_organization.html#mix-and-match-configuration-formats) is not applicable for our use case. We need something more in the line of the `$ref` method of json-schema.

**Comment by Doqnach** [comment]
A bit scary work-around:

https://github.com/Doqnach/miraizou-core/blob/master/src/ModifyParentPrivates.php
https://github.com/Doqnach/miraizou-yaml/blob/master/src/YamlParser.php

**Comment by stof** [comment]
I don't think that's something we want to implement, as that's not how YAML references are meant to work. YAML references are resolved inside the YAML document.

**Comment by Doqnach** [comment]
That, I surely understand. YAML itself of course does not support 'sub documents'/includes, so references are by definition only applied to the (main) document itself.

Another approach would be to be able to give the parser handlers for custom tags that are bound to the context (`$this`) of the parser? Because now all custom tag handling can only be done after the parser has finished (in my case an `array_walk_recursive()`  looking for any array item with a `TaggedValue()` instance). I would love to be able to hook into the actual behaviour of the parser when it hits a custom tag so I have access to the state of the parser (and thus it's `$this->refs[]`).

**Comment by fabpot** [comment]
I'm closing as we won't implement things that are out of the scope of the YAML spec.

**Comment by ToBe998** [comment]
This is actually not a good reason. The custom include tag is loading content into the one YAML file. So for the yaml partser it is all within the one yaml context and part of the one file.

This is actually a very important feature as it allows complex yaml data to be abstracted and seperated. This is such an important feature of basically any design pattern in use that this feature should actually be part of symfony/yaml itself (and was in earlier versions, at least indirectly through the old congfig package).

Long story short. This is completely within the yaml spec.

**Comment by stof** [comment]
@ToBe998 no, the custom include tag is something processed **after** parsing the YAML, but the code receiving the `TaggedValue` object returned by the parser.

If you want something that actually bring content inside the YAML before parsing, it means your file is not actually a YAML file but a template that gets rendered/processed by another tool to generate YAML (and that's also something that does not need any support of the YAML parser itself as it is about changing the input of the parser.

**Comment by ToBe998** [comment]
@stof If you add something **after** parsing YAML, yo need to duplicate any feature that is already present in YAML. For example references. And i think we can all agree that this would be a very bad idea. So, i can understand that adding this feature is complex and maybe out of scope at this time, but it is most certainly not unecessary.
Look at other parsers like php_yaml. All of them work inside the YAML parser and not after it because that makes it possible to let such custom functions work in combination and not only in their own context.

**Comment by stof** [comment]
which parser are you talking about ?
